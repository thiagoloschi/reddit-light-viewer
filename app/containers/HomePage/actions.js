import { fetchPosts as getPosts } from 'services/reddit';
import { FETCH_POSTS } from './constants';

export function fetchPosts(topic, sort, after) {
  return {
    type: FETCH_POSTS,
    promise: getPosts(topic, sort, after),
  };
}
