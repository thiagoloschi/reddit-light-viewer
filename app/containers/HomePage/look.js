import styled from 'styled-components';

export const FilterSection = styled.section`
  display: flex;
  align-items: baseline;
`;
