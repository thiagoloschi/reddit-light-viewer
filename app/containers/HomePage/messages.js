import { defineMessages } from 'react-intl';

export const scope = 'app.containers.HomePage';

export default defineMessages({
  searchbarPlaceholder: {
    id: `${scope}.searchbarPlaceholder`,
    defaultMessage: 'look for a topic!',
  },
});
