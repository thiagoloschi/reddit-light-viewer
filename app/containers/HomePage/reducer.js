import { handle } from 'redux-pack';
import { fromJS } from 'immutable';
import { FETCH_POSTS } from './constants';

export const initialState = fromJS({
  isLoading: false,
  error: null,
  posts: [],
});

function homeReducer(state = initialState, action) {
  const { type, payload } = action;
  switch (type) {
    case FETCH_POSTS:
      return handle(state, action, {
        start: prevState => prevState.merge({ isLoading: true }),
        finish: prevState => prevState.merge({ isLoading: false }),
        failure: prevState => prevState.merge({ error: payload }),
        success: prevState => prevState.merge({ posts: payload.data }),
      });
    default:
      return state;
  }
}

export default homeReducer;
