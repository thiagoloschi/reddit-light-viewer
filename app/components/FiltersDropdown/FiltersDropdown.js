import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from './look';

function FiltersDropdown({ filters, selectedFilter, onChange }) {
  if (!filters || filters.length <= 0) {
    return (
      <Dropdown onChange={onChange}>
        <option> - </option>
      </Dropdown>
    );
  }

  return (
    <Dropdown onChange={onChange}>
      {filters.map(topic => (
        <option
          key={topic}
          value={topic}
          defaultValue={topic === selectedFilter}
        >
          {topic}
        </option>
      ))}
    </Dropdown>
  );
}

FiltersDropdown.propTypes = {
  filters: PropTypes.array,
  selectedFilter: PropTypes.string,
  onChange: PropTypes.func,
};

export default FiltersDropdown;
