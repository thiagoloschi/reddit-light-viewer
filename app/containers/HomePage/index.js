/* 
  Following exception is added in order to make hot reloading possible as it doesn't support stateless functional
  components
*/
/* eslint-disable react/prefer-stateless-function */
import React from 'react';
import { FormattedMessage } from 'react-intl';
import messages from './messages';

export default class HomePage extends React.PureComponent {
  render() {
    return (
      <h1>
        <FormattedMessage {...messages.header} />
      </h1>
    );
  }
}
