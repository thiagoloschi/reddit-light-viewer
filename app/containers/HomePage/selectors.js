import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectHome = state => state.get('home', initialState);

const makeSelectPosts = () =>
  createSelector(selectHome, homeState => homeState.get('posts').toJS());

const makeSelectError = () =>
  createSelector(selectHome, homeState => homeState.get('error'));

const makeSelectIsLoading = () =>
  createSelector(selectHome, homeState => homeState.get('isLoading'));

export { selectHome, makeSelectPosts, makeSelectError, makeSelectIsLoading };
