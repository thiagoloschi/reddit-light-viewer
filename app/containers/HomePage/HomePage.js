/* 
  Following exception is added in order to make hot reloading possible as it doesn't support stateless functional
  components
*/
/* eslint-disable react/prefer-stateless-function */
import React from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';

import Posts from 'components/PostFactory';
import TopicDropdown from 'components/TopicDropdown';

import { fetchPosts } from './actions';
import { makeSelectPosts } from './selectors';
import reducer from './reducer';

class HomePage extends React.PureComponent {
  static propTypes = {
    posts: PropTypes.object,
    getPosts: PropTypes.func,
  };

  constructor(props) {
    super(props);
    this.fetchNewTopic = this.fetchNewTopic.bind(this);
    this.fetchMorePosts = this.fetchMorePosts.bind(this);
    this.state = {
      topics: ['hot', 'new', 'controversial', 'top', 'rising'],
      selectedTopic: 'hot',
    };
  }

  componentDidMount() {
    const {
      posts: { children },
      getPosts,
    } = this.props;

    if (!children || children.length <= 0) {
      getPosts();
    }

    window.addEventListener('scroll', this.fetchMorePosts);
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.fetchMorePosts);
  }

  fetchNewTopic({ target: { value } }) {
    this.props.getPosts(value);

    this.setState({
      selectedTopic: value,
    });
  }

  fetchMorePosts() {
    const {
      posts: { after },
      getPosts,
    } = this.props;
    const { selectedTopic } = this.state;
    const { scrollTop, scrollHeight, offsetHeight } = document.documentElement;
    if (scrollTop + offsetHeight === scrollHeight) {
      getPosts(selectedTopic, after);
    }
  }

  render() {
    const {
      posts: { children },
    } = this.props;
    const { topics, selectedTopic } = this.state;

    return (
      <>
        <TopicDropdown
          selectedTopic={selectedTopic}
          topics={topics}
          onChange={this.fetchNewTopic}
        />
        {children && children.length > 0 && <Posts posts={children} />}
      </>
    );
  }
}

const mapStateToProps = createStructuredSelector({
  posts: makeSelectPosts(),
});

export function mapDispatchToProps(dispatch) {
  return {
    getPosts: (query, after) => dispatch(fetchPosts(query, after)),
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withReducer = injectReducer({ key: 'home', reducer });

export default compose(
  withReducer,
  withConnect,
)(HomePage);
