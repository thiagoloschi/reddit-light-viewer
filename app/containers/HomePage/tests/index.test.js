import React from 'react';
import { shallow } from 'enzyme';
import Posts from 'components/PostFactory';
import FiltersDropdown from 'components/FiltersDropdown';
import SearchBar from 'components/SearchBar';
import mockedPost from 'utils/mockedPost.json';

import { HomePage, mapDispatchToProps } from '../HomePage';
import { NavigationBar, FilterSection, Heading } from '../look';

import { fetchPosts } from '../actions';

jest.mock('services/reddit');

const defaultProps = {
  error: false,
  isLoading: false,
  getPosts: jest.fn(),
  posts: {
    after: 'after',
    children: [mockedPost],
  },
};

const defaultState = {
  filters: ['hot', 'new', 'controversial', 'top', 'rising'],
  selectedFilter: 'hot',
  topic: 'all',
};

const render = props => shallow(<HomePage {...defaultProps} {...props} />);

describe('<HomePage />', () => {
  const { getPosts } = defaultProps;

  describe('componentDidMount', () => {
    describe('when there is no error and no posts', () => {
      const withNoErrorAndNoPosts = render({ posts: [], error: false });
      it('should call getPosts when the component mounts', () => {
        withNoErrorAndNoPosts.instance().componentDidMount();
        expect(getPosts).toHaveBeenCalled();
      });
    });

    it('should add a listener to onsroll', () => {
      const withListener = render({ posts: [mockedPost, mockedPost] });
      window.addEventListener = jest.fn();
      withListener.instance().componentDidMount();
      const callback = withListener.instance().fetchMorePosts;
      expect(window.addEventListener).toHaveBeenCalledWith('scroll', callback);
    });
  });

  describe('componentWillUnmount', () => {
    it('should remove the listener to onsroll', () => {
      const withoutListener = render({ posts: [mockedPost] });
      window.removeEventListener = jest.fn();
      withoutListener.instance().componentWillUnmount();
      const callback = withoutListener.instance().fetchMorePosts;
      expect(window.addEventListener).toHaveBeenCalledWith('scroll', callback);
    });
  });

  describe('lookForATopic', () => {
    const homePage = render();
    const event = {
      preventDefault: jest.fn(),
    };
    homePage.instance().state.topic = 'notEmpty';
    homePage.instance().state.selectedFilter = 'someFilter';
    homePage.instance().lookForATopic(event);
    it('should call event.preventdefault', () => {
      expect(event.preventDefault).toHaveBeenCalled();
    });

    it('should call getPosts if topic is not empty', () => {
      expect(getPosts).toHaveBeenCalledWith('notEmpty', 'someFilter');
    });
  });

  describe('updateTopic', () => {
    it('should update the state with the right value', () => {
      const homePage = render();
      const event = { target: { value: 'newValue' } };
      homePage.instance().updateTopic(event);
      expect(homePage.instance().state.topic).toEqual('newValue');
    });
  });

  describe('fetchWithFilter', () => {
    const homePage = render();
    const event = { target: { value: 'someFilter' } };
    homePage.instance().state.topic = 'notEmpty';
    homePage.instance().fetchWithFilter(event);

    it('should update the stat with the selected filter', () => {
      expect(homePage.instance().state.selectedFilter).toEqual('someFilter');
    });

    it('should call getPosts if topic is not empty', () => {
      expect(getPosts).toHaveBeenCalledWith('notEmpty', 'someFilter');
    });
  });

  describe('fetchMorePosts', () => {
    const homePage = render();
    homePage.instance().state.topic = 'notEmpty';
    homePage.instance().state.selectedFilter = 'selectedFilter';
    const { scrollHeight } = document.documentElement;
    homePage.simulate('scroll', { deltaY: scrollHeight });

    describe('when there are more posts and it is the bottom of the page', () => {
      it('should call getPosts if topic is not empty', () => {
        expect(getPosts).toHaveBeenCalledWith('notEmpty', 'someFilter');
      });
    });
  });

  describe('render', () => {
    const homePage = render();
    const {
      posts: { children },
      error,
      isLoading,
    } = defaultProps;
    const { filters, selectedFilter, topic } = defaultState;
    const { lookForATopic, updateTopic, fetchWithFilter } = homePage.instance();
    it('should render correctly with the matching props', () => {
      expect(
        homePage.containsMatchingElement(
          <>
            <NavigationBar>
              <Heading>{`r/${topic}`}</Heading>
              <FilterSection>
                <SearchBar
                  name="search"
                  onSubmit={lookForATopic}
                  onChange={updateTopic}
                  placeholder="look for a topic"
                />
                <FiltersDropdown
                  selectedFilter={selectedFilter}
                  filters={filters}
                  onChange={fetchWithFilter}
                />
              </FilterSection>
            </NavigationBar>
            <Posts error={error} isLoading={isLoading} posts={children} />
          </>,
        ),
      ).toEqual(true);
      expect(homePage).toMatchSnapshot();
    });
  });

  describe('mapDispatchToProps', () => {
    const dispatch = jest.fn();
    const dispatchProps = mapDispatchToProps(dispatch);

    describe('dispatchProps', () => {
      it('should have dispatchProps defined', () => {
        expect(dispatchProps.getPosts).toBeDefined();
      });
    });

    describe('getPosts', () => {
      it('should dispatch fetchPosts correctly', () => {
        dispatchProps.getPosts();
        expect(dispatch).toHaveBeenCalledWith(fetchPosts());
      });
    });
  });
});
