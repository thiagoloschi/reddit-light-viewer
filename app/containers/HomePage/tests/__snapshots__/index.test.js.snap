// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> render should render correctly with the matching props 1`] = `
<Fragment>
  <look__NavigationBar>
    <look__Heading>
      r/all
    </look__Heading>
    <look__FilterSection>
      <SearchBar
        name="search"
        onChange={[Function]}
        onSubmit={[Function]}
        placeholder="look for a topic"
      />
      <FiltersDropdown
        filters={
          Array [
            "hot",
            "new",
            "controversial",
            "top",
            "rising",
          ]
        }
        onChange={[Function]}
        selectedFilter="hot"
      />
    </look__FilterSection>
  </look__NavigationBar>
  <PostFactory
    error={false}
    isLoading={false}
    posts={
      Array [
        Object {
          "data": Object {
            "author": "namraka",
            "created": 1547059800,
            "created_utc": 1547031000,
            "id": "ae5q0v",
            "num_comments": 1290,
            "num_crossposts": 42,
            "permalink": "/r/gifs/comments/ae5q0v/nobody_noticed_just_act_normal/",
            "score": 90494,
            "subreddit": "gifs",
            "subreddit_id": "t5_2qt55",
            "subreddit_name_prefixed": "r/gifs",
            "thumbnail": "https://b.thumbs.redditmedia.com/4X6iAiWpar-2hfQUaPhLLhzGBEd4NW2X6piBlOkGkJA.jpg",
            "thumbnail_height": 140,
            "thumbnail_width": 140,
            "title": "Nobody noticed, just act normal",
            "ups": 90494,
            "url": "https://v.redd.it/je38z1kund921",
          },
        },
      ]
    }
  />
</Fragment>
`;
